polling_stations/apps/data_importers/data_types.py

```diff
@@ -0,0 +1,192 @@
+            )
+            return
+
+        self.elements.append(address)
+
+    def get_uprn_lookup(self):
+        # for each address, build a lookup of uprn -> set of station ids
+        uprn_lookup = {}
+        for record in self.elements:
+            uprn = record["uprn"]
+            if not uprn:
+                continue
+            if uprn in uprn_lookup:
+                uprn_lookup[uprn].add(record["polling_station_id"])
+            else:
+                uprn_lookup[uprn] = set([record["polling_station_id"]])
+
+        return uprn_lookup
+
+    # TODO be more clever to report on duplicates.
+    def remove_duplicate_uprns(self):
+        uprn_lookup = self.get_uprn_lookup()
+        self.elements = [
+            record for record in self.elements if len(uprn_lookup[record["uprn"]]) == 1
+        ]
+
+    @property
+    def council_id(self):
+        for e in self.elements:
+            return e["council"].council_id
+
+    @property
+    def council_ids(self):
+        # hack to deal with new local auths
+        # some councils won't exist in all the data yet
+        if self.council_id in settings.NEW_COUNCILS:
+            return tuple(
+                [k for k, v in settings.OLD_TO_NEW_MAP.items() if v == self.council_id]
+            )
+        else:
+            return (self.council_id,)
+
+    def get_polling_station_lookup(self):
+        # for each address, build a lookup of polling_station_id -> set of uprns
+        polling_station_lookup = {}
+        for record in self.elements:
+            if record["polling_station_id"] in polling_station_lookup:
+                polling_station_lookup[record["polling_station_id"]].add(record["uprn"])
+            else:
+                polling_station_lookup[record["polling_station_id"]] = set(
+                    record["uprn"]
+                )
+
+        return polling_station_lookup
+
+    def remove_records_not_in_addressbase(self, addressbase_data):
+        self.elements = [e for e in self.elements if e["uprn"] in addressbase_data]
+
+    def remove_records_that_dont_match_addressbase(
+        self, addressbase_data, match_threshold
+    ):
+        for input_record in self.elements:
+            addressbase_record = addressbase_data[input_record["uprn"]]
+            match_quality = fuzz.partial_ratio(
+                input_record["address"].lower().replace(",", ""),
+                addressbase_record["address"].lower().replace(",", ""),
+            )
+            if (
+                Postcode(input_record["postcode"]).with_space
+                == Postcode(addressbase_record["postcode"]).with_space
+                and match_quality >= 0.9 * match_threshold
+            ):
+                continue
+            elif (
+                Postcode(input_record["postcode"]).with_space
+                != Postcode(addressbase_record["postcode"]).with_space
+                and match_quality >= 0.9 * match_threshold
+            ):
+                continue
```

Comment:
I suspect this is something you're probably currently tweaking, but with this block here:

```py
if (
    Postcode(input_record["postcode"]).with_space
    == Postcode(addressbase_record["postcode"]).with_space
    and match_quality >= 0.9 * match_threshold
):
    continue
elif (
    Postcode(input_record["postcode"]).with_space
    != Postcode(addressbase_record["postcode"]).with_space
    and match_quality >= 0.9 * match_threshold
):
    continue
```

As written, isn't this just the same as just saying

```py
if (match_quality >= 0.9 * match_threshold):
    continue
```

?

i.e: we're completely ignoring postcode and _only_ looking at address fuzzy match.

Also, why `0.9 * match_threshold`? That makes `match_threshold` a bit counter-intuitive because setting it to 100 means 90%. Is that just the values you were last playing with?

Reply:
:slightly_frowning_face: Yep that's me in the middle of tweaking it. 
```
if (
    Postcode(input_record["postcode"]).with_space
    == Postcode(addressbase_record["postcode"]).with_space
    and match_quality >= 0.9 * match_threshold
):
    continue
elif (
    Postcode(input_record["postcode"]).with_space
    != Postcode(addressbase_record["postcode"]).with_space
    and match_quality >=  match_threshold
):
    continue
```
Was my intention. I'm now not sure which one I made the stats with, so I'll check that.